xarray/core/pycompat.py

```diff
@@ -101,3 +101,37 @@ def is_chunked_array(x) -> bool:
 
 def is_0d_dask_array(x):
     return is_duck_dask_array(x) and is_scalar(x)
+
+
+def to_numpy(data):
+    from xarray.core.indexing import ExplicitlyIndexed
+    from xarray.core.parallelcompat import get_chunked_array_type
+
+    if isinstance(data, ExplicitlyIndexed):
+        data = data.get_duck_array()
+
+    # TODO first attempt to call .to_numpy() once some libraries implement it
+    if hasattr(data, "chunks"):
+        chunkmanager = get_chunked_array_type(data)
+        data, *_ = chunkmanager.compute(data)
+    if isinstance(data, array_type("cupy")):
+        data = data.get()
+    # pint has to be imported dynamically as pint imports xarray
+    if isinstance(data, array_type("pint")):
+        data = data.magnitude
+    if isinstance(data, array_type("sparse")):
+        data = data.todense()
+    data = np.asarray(data)
+
+    return data
+
+
+def to_duck_array(data):
```

Comment:
Some typing would be nice, but again no idea what the return type is. Do we have a duck array protocol?

I think for internal use a protocol is nice, for external return types we should simply return Any.